src/tools/rbd_nbd/rbd-nbd.cc

```diff
@@ -792,34 +794,106 @@ static int parse_imgpath(const std::string &imgpath)
   return 0;
 }
 
-static int do_list_mapped_devices()
+static void get_mapped_info(int pid, std::string &pool_name, std::string &image_name, std::string &snap_name)
 {
-  char path[64];
-  int m = 0;
-  int fd[2];
+  std::string path = "/proc/" + stringify(pid) + "/cmdline";
+  std::ifstream ifs;
+  std::string cmdline;
 
-  if (socketpair(AF_UNIX, SOCK_STREAM, 0, fd) == -1) {
-    int r = -errno;
-    cerr << "rbd-nbd: socketpair failed: " << cpp_strerror(-r) << std::endl;
-    return r;
+  ifs.open(path.c_str(), std::ifstream::in);
+  assert (ifs.is_open());
+  ifs >> cmdline;
+
+  for(unsigned int index = 0; index < cmdline.size(); index++) {
+    if(((unsigned char)cmdline[index]) < ' ') {
+      cmdline[index] = ' ';
+      index++;
+    }
+  }
+
+  char *pch = const_cast<char*>(cmdline.c_str());
+  pch = strtok(pch, " ");
+  while (pch != NULL) {
+    if (strcmp(pch, "map") == 0) {
+      pch = strtok(NULL, " ");
+      while (pch != NULL) {
+        if (strlen(pch) > 2 && pch[0] == '-' && pch[1] == '-') {
+          pch = strtok(NULL, " ");
+        } else {
+          break;
+        }
+      }
+      if (pch != NULL) {
+        break;
+      }
+    } else {
+      pch = strtok(NULL, " ");
+    }
+  }
+
+  image_name = strtok(pch, "/");
+  pch = strtok(NULL, "/");
+  if (pch == NULL) {
+    pool_name = "rbd";
+  } else {
+    pool_name = image_name;
+    image_name = pch;
+  }
+
+  pch = const_cast<char*>(image_name.c_str());
+  image_name = strtok(pch, "@");
+  pch = strtok(NULL, "@");
+  if (pch != NULL) {
+    snap_name = pch;
+  }
+}
+
+static int get_map_pid(const std::string& pid_path)
+{
+  int pid = 0;
+  std::ifstream ifs;
+  ifs.open(pid_path.c_str(), std::ifstream::in);
+  if (!ifs.is_open()) {
+    return 0;
   }
+  ifs >> pid;
+  return pid;
+}
+
+static int do_list_mapped_devices()
+{
+  int index = 0;
+  int id = 0;
+  int pid = 0;
+  TextTable tbl;
+
+  tbl.define_column("id", TextTable::LEFT, TextTable::LEFT);
+  tbl.define_column("device", TextTable::LEFT, TextTable::LEFT);
+  tbl.define_column("pool", TextTable::LEFT, TextTable::LEFT);
+  tbl.define_column("image", TextTable::LEFT, TextTable::LEFT);
+  tbl.define_column("snap", TextTable::LEFT, TextTable::LEFT);
 
   while (true) {
-    snprintf(path, sizeof(path), "/dev/nbd%d", m);
-    int nbd = open_device(path);
-    if (nbd < 0)
+    std::string nbd_path = "/sys/block/nbd" + stringify(index);
+    if(access(nbd_path.c_str(), F_OK) != 0) {
       break;
-    if (ioctl(nbd, NBD_SET_SOCK, fd[0]) != 0)
-      cout << path << std::endl;
-    else
-      ioctl(nbd, NBD_CLEAR_SOCK);
-    close(nbd);
-    m++;
-  }
+    }
+    std::string pid_path = nbd_path + "/pid";
+    pid = get_map_pid(pid_path);
+
+    if(pid > 0) {
+      std::string pool_name, image_name, snap_name;
+      get_mapped_info(pid, pool_name, image_name, snap_name);
```

Comment:
Probably a good idea for `get_mapped_info` to return a error code (whether it succeeded) so we could filter out nbd devices created by other (not rbd-nbd) processes.